In [ ]:
from google.colab import drive
import pandas as pd

# Read the CSV file from drive
drive.mount('/content/drive')
df = pd.read_csv('/content/drive/MyDrive/Airbnb_Data/airnb.csv')
#df2 = pd.read_csv('/content/drive/MyDrive/Airbnb_Data/airnb_desert.csv', encoding='latin1')
#df3 = pd.read_csv('/content/drive/MyDrive/Airbnb_Data/airnb_luxe.csv', encoding='latin1')

display(df.head())
#display(df2.head())
#display(df3.head())

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed


In [ ]:
#1. # Checking for Null Values
df.isnull().sample(30)

,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
340,False,False,False,False,True,False,False
693,False,False,False,False,True,False,False
250,False,False,False,False,True,False,False
682,False,False,False,False,True,False,False
130,False,False,False,False,False,False,False
655,False,False,False,False,True,False,False
846,False,False,False,False,True,False,False
453,False,False,False,False,True,False,False
930,False,False,False,False,True,False,False
939,False,False,False,False,True,False,False


In [ ]:
# Check for null values in each column
df.isna().any()

,0
Title,False
Detail,False
Date,False
Price(in dollar),False
Offer price(in dollar),True
Review and rating,True
Number of bed,False


In [ ]:
# Count the number of null values in each column
df.isna().sum()

,0
Title,0
Detail,0
Date,0
Price(in dollar),0
Offer price(in dollar),787
Review and rating,6
Number of bed,0


In [ ]:
#Null values in each row
df.isna().sum(axis=1).sample(20)

,0
461,1
805,1
222,1
529,1
169,1
698,1
216,0
90,0
407,1
853,1


In [ ]:
df.shape

(953, 7)

In [ ]:
df.info()
df.describe(include="all")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Title                   953 non-null    object
 1   Detail                  953 non-null    object
 2   Date                    953 non-null    object
 3   Price(in dollar)        953 non-null    object
 4   Offer price(in dollar)  166 non-null    object
 5   Review and rating       947 non-null    object
 6   Number of bed           953 non-null    object
dtypes: object(7)
memory usage: 52.2+ KB


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
count,953,953,953,953,166,947,953
unique,756,839,204,322,108,754,33
top,"Apartment in Kuala Lumpur, Malaysia",Cozy room 48.5 sq.m @The Astra near Night market,May 1 - 6,111.00,63.00,New,2 beds
freq,7,3,219,13,4,16,221


In [ ]:
# Dropping rows with Null Values
df.dropna()  # Displays how it will look the DataFrame without the rows with null values

,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed
...,...,...,...,...,...,...,...
828,Apartment in Houston,King Suite/Sky Lounge/Luxurious Experience.,May 1 - 6,108.00,148.00,4.85 (55),2 beds
829,Cabin in Sevierville,Few days left in May,May 1 - 6,286.00,112.00,4.98 (263),1 king bed
830,Cabin in Fountain Green,"Cozy Private Mountain Cabin on 1,000 acre prop...",May 1 - 6,186.00,103.00,4.96 (78),4 beds
882,Dome in El Prado,Geodesic Earth Dome,May 1 - 6,91.00,161.00,4.9 (953),2 beds


### #This finding means that if we drop all rows with Null Values, the Data Frame will loose sustancial information and the rows will be reduced from 953 to 166

In [ ]:
# Dropping columns with too many Null Values
df.dropna(axis=1)
#Aquí se perderí información porque desaparecerían la Columna Offer Price y Raiting Reviews

,Title,Detail,Date,Price(in dollar),Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,1 queen bed
...,...,...,...,...,...
948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,1 king bed
949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,1 king bed
950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,5 beds
951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,4 beds


In [ ]:
#Checking Data Types
df.dtypes


,0
Title,object
Detail,object
Date,object
Price(in dollar),object
Offer price(in dollar),object
Review and rating,object
Number of bed,object


In [ ]:
#Dealing with duplicates
df.duplicated().sum()

np.int64(34)

In [ ]:
# Check for duplicates in specific columns
df.duplicated(subset=['Title']).sum()

np.int64(197)

In [ ]:
# Check for duplicates in specific columns
df.duplicated(subset=['Detail']).sum()

np.int64(114)

In [ ]:
# Check for duplicates in 3 specific columns at the same time to fin especific duplicates
df.duplicated(subset=['Title','Detail','Date']).sum()

np.int64(103)

In [ ]:
#Droping duplicates with the same Title, Detail and Date
df.drop_duplicates(subset=['Title','Detail','Date'])

,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed
...,...,...,...,...,...,...,...
948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed
949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed
950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds
951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds


In [ ]:
#To leave the 850 rows without duplicates we drop the duplicates found with the three columns Title, Detail and Date
df.drop_duplicates(subset=['Title','Detail','Date'], inplace = True)


In [ ]:
df_without_duplicates = df.copy()
df_without_duplicates.sample(20)

,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
554,"Home in Kecamatan Ubud, Indonesia",PROMO! Luxury Villa Escape in Ubud 2 BR,Sep 1 - 6,306.00,NaN,5.0 (4),2 beds
116,"Shipping container in Playa Hermosa, Costa Rica",Beachfront Bungalow w/ private hydromassage po...,Jun 10 - 15,147.00,NaN,4.9 (68),2 beds
788,Cottage in Crozet,Idyllic Cottage Retreat,May 1 - 6,188.00,NaN,4.86 (14),2 beds
688,"Villa in Uluwatu, Indonesia","Spectacular Ocean View Villa in Uluwatu, Bali",Jun 14 - 19,673.00,NaN,4.94 (143),2 beds
80,"Condo in Lahaina, Hawaii, US",Noelani-My Perfect Stays Special: Jun 9-16 $189,Jun 9 - 14,302.00,NaN,4.82 (71),1 king bed
754,Home in Eustis,Lake House Minutes to Downtown Mt. Dora and Eu...,May 1 - 6,173.00,NaN,4.99 (179),2 beds
757,Cabin in Moab,Basic Cabin #1 - Sleeps 4 - OK-C1,May 6 - 11,125.00,NaN,4.86 (233),3 beds
97,"Cabin in Zapotal, Costa Rica",Pipe’s Lodge,Jun 4 - 9,51.00,NaN,4.87 (31),1 double bed
111,"Treehouse in Wardensville, West Virginia, US",Eagles Nest Treehouse w/Hot tub!,Jun 4 - 10,244.00,63.00,New,3 beds
438,"Chalet in Covilhã, Portugal",Xitaca do Pula,Jun 11 - 17,78.00,NaN,4.95 (203),2 king beds


In [ ]:
df_without_duplicates.reset_index().tail()

,index,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
845,948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed
846,949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed
847,950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds
848,951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds
849,952,Room in Long Beach,Across the Street from Beach Getaway/Near Conv...,May 1 - 6,106.00,NaN,4.85 (97),4 beds


In [ ]:
df_without_duplicates.info()

<class 'pandas.core.frame.DataFrame'>
Index: 850 entries, 0 to 952
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Title                   850 non-null    object
 1   Detail                  850 non-null    object
 2   Date                    850 non-null    object
 3   Price(in dollar)        850 non-null    object
 4   Offer price(in dollar)  131 non-null    object
 5   Review and rating       844 non-null    object
 6   Number of bed           850 non-null    object
dtypes: object(7)
memory usage: 53.1+ KB


## Cleaning Columns

In [ ]:
df_without_duplicates.columns

Index(['Title', 'Detail', 'Date', 'Price(in dollar)', 'Offer price(in dollar)',
       'Review and rating', 'Number of bed'],
      dtype='object')

In [ ]:
#Just changing names of the colums
df_without_duplicates.columns = [(column.lower()).replace(" ","_") for column in df_without_duplicates.columns]
df_without_duplicates.columns

Index(['title', 'detail', 'date', 'price(in_dollar)', 'offer_price(in_dollar)',
       'review_and_rating', 'number_of_bed'],
      dtype='object')

In [ ]:
df_without_duplicates.head()

,title,detail,date,price(in_dollar),offer_price(in_dollar),review_and_rating,number_of_bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed


In [ ]:
#Clean unnecessary columns
df_without_duplicates.drop('detail',axis=1,inplace =True)
df_without_duplicates.drop('date',axis=1,inplace =True)
df_without_duplicates.drop('number_of_bed',axis=1,inplace =True)
df_without_duplicates.drop('offer_price(in_dollar)',axis=1,inplace =True)

In [ ]:
df_without_duplicates.columns

Index(['title', 'price(in_dollar)', 'review_and_rating'], dtype='object')

Data Frame Clean After:


*   1. Evaluating Null Values
*   2. Identifiying and dropping duplicates
*   3. Droping unnecessary columns for this study





In [ ]:
#Data Frame clean but still needing to separate information
df_clean = df_without_duplicates.copy()
df_clean.sample(10)

,title,price(in_dollar),review_and_rating
152,"Camper/RV in Ashford, Washington, US",286.00,5.0 (11)
164,"Apartment in Cancún, Mexico",45.00,4.88 (306)
577,"Apartment in Bangkok, Thailand",19.00,4.95 (438)
909,Apartment in Miami Beach,153.00,4.56 (418)
726,Home in Fairlee,244.00,4.95 (344)
323,"Camper/RV in Mayen, Germany",48.00,NaN
920,Room in New York,106.00,4.9 (127)
412,"Cabin in Bridge of Earn, UK",64.00,4.67 (9)
736,Cabin in Crivitz,149.00,4.99 (72)
899,Apartment in Houston,113.00,4.84 (144)


In [ ]:
display(df_clean.sample(10))

,title,price(in_dollar),review_and_rating
720,Cabin in Branson,129.00,4.86 (22)
534,"Home in Tegallalang, Indonesia",33.00,4.97 (229)
116,"Shipping container in Playa Hermosa, Costa Rica",147.00,4.9 (68)
97,"Cabin in Zapotal, Costa Rica",51.00,4.87 (31)
872,Guesthouse in Colorado Springs,115.00,4.96 (170)
452,"Hut in Ubud, Indonesia",171.00,4.79 (164)
649,"Hotel in Nha Trang, Vietnam",17.00,4.35 (17)
947,Condo in Longboat Key,196.00,4.89 (415)
88,"Place to stay in Punta Cana, Dominican Republic",132.00,4.86 (29)
885,Condo in Saint Petersburg,143.00,4.87 (360)


NEXT STEP: This allow us to identify that the Column 'title' actualy contens information about the location but it needs to be extracted, as well as separate rating and reviews


In [ ]:
import pandas as pd

def title_extractor(df: pd.DataFrame) -> pd.DataFrame:
    temp_df = df.copy()
    temp_df[["header", "state", "country"]] = temp_df["title"].str.split(
        ",", expand=True
    )
    temp_df.drop("title", inplace=True, axis=1)
    return temp_df

In [ ]:
df_extracted_title = title_extractor(df_clean)
display(df_extracted_title.head(10))

,price(in_dollar),review_and_rating,header,state,country
0,306.00,4.85 (531),Chalet in Skykomish,Washington,US
1,485.00,4.77 (146),Cabin in Hancock,New York,US
2,119.00,4.91 (515),Cabin in West Farmington,Ohio,US
3,192.00,4.94 (88),Home in Blue Ridge,Georgia,US
4,232.00,4.99 (222),Treehouse in Grandview,Texas,US
5,261.00,4.84 (555),Tiny home in Puerto Escondido,Mexico,None
6,607.00,4.64 (90),Home in Saco,Maine,US
7,381.00,5.0 (48),Chalet in Grand Étang,Canada,None
8,29.00,4.84 (58),Room in Mexico City,Mexico,None
9,154.00,4.92 (744),Tiny home in Sandy Valley,Nevada,US


In [ ]:
df_extracted_title.columns = [(column.lower()).replace(" ","_") for column in df_extracted_title.columns]
display(df_extracted_title)

,price(in_dollar),review_and_rating,header,state,country
0,306.00,4.85 (531),Chalet in Skykomish,Washington,US
1,485.00,4.77 (146),Cabin in Hancock,New York,US
2,119.00,4.91 (515),Cabin in West Farmington,Ohio,US
3,192.00,4.94 (88),Home in Blue Ridge,Georgia,US
4,232.00,4.99 (222),Treehouse in Grandview,Texas,US
...,...,...,...,...,...
948,141.00,4.85 (82),Guest suite in Lake Worth,None,None
949,43.00,4.9 (300),Room in Kissimmee,None,None
950,73.00,4.92 (282),Guesthouse in San Antonio,None,None
951,233.00,4.96 (318),Apartment in Englewood,None,None


In [ ]:
#Once we have extracted the title we need to separate the information in the review_and_rating column

In [ ]:
df_extracted_title[['raiting','reviews']] = df_extracted_title.review_and_rating.str.split(" ", expand=True)
display(df_extracted_title.head(10))

,price(in_dollar),review_and_rating,header,state,country,raiting,reviews
0,306.00,4.85 (531),Chalet in Skykomish,Washington,US,4.85,(531)
1,485.00,4.77 (146),Cabin in Hancock,New York,US,4.77,(146)
2,119.00,4.91 (515),Cabin in West Farmington,Ohio,US,4.91,(515)
3,192.00,4.94 (88),Home in Blue Ridge,Georgia,US,4.94,(88)
4,232.00,4.99 (222),Treehouse in Grandview,Texas,US,4.99,(222)
5,261.00,4.84 (555),Tiny home in Puerto Escondido,Mexico,None,4.84,(555)
6,607.00,4.64 (90),Home in Saco,Maine,US,4.64,(90)
7,381.00,5.0 (48),Chalet in Grand Étang,Canada,None,5.0,(48)
8,29.00,4.84 (58),Room in Mexico City,Mexico,None,4.84,(58)
9,154.00,4.92 (744),Tiny home in Sandy Valley,Nevada,US,4.92,(744)


In [ ]:
df_extracted_title["reviews"] = df_extracted_title["reviews"].str.replace(r'[()]', '', regex=True)
display(df_extracted_title.sample(9))

,price(in_dollar),review_and_rating,header,state,country,raiting,reviews
940,124.00,4.7 (242),Condo in Las Vegas,None,None,4.7,242
75,89.00,4.78 (231),Apartment in Playa del Carmen,Mexico,None,4.78,231
331,184.00,4.98 (46),Vacation home in Akrotiri,Greece,None,4.98,46
764,151.00,4.93 (555),Cottage in Los Angeles,None,None,4.93,555
555,130.00,4.79 (112),Apartment in Candidasa,Indonesia,None,4.79,112
62,55.00,4.6 (73),Apartment in Tulum,Mexico,None,4.6,73
951,233.00,4.96 (318),Apartment in Englewood,None,None,4.96,318
483,34.00,5.0 (37),Room in Kecamatan Denpasar Selatan,Indonesia,None,5.0,37
700,49.00,4.89 (37),Place to stay in Kamala,Thailand,None,4.89,37


In [ ]:
df_extracted_title.drop('review_and_rating',axis=1,inplace =True)
df_extracted_title.columns

Index(['price(in_dollar)', 'header', 'state', 'country', 'raiting', 'reviews'], dtype='object')

In [ ]:
df_new=df_extracted_title.copy()
display(df_new.head(10))

,price(in_dollar),header,state,country,raiting,reviews
0,306.00,Chalet in Skykomish,Washington,US,4.85,531
1,485.00,Cabin in Hancock,New York,US,4.77,146
2,119.00,Cabin in West Farmington,Ohio,US,4.91,515
3,192.00,Home in Blue Ridge,Georgia,US,4.94,88
4,232.00,Treehouse in Grandview,Texas,US,4.99,222
5,261.00,Tiny home in Puerto Escondido,Mexico,None,4.84,555
6,607.00,Home in Saco,Maine,US,4.64,90
7,381.00,Chalet in Grand Étang,Canada,None,5.0,48
8,29.00,Room in Mexico City,Mexico,None,4.84,58
9,154.00,Tiny home in Sandy Valley,Nevada,US,4.92,744


In [ ]:
#Since we are going to need the values of rating, review and price as number I convert those columns to number

In [ ]:
df_new["price(in_dollar)"] = pd.to_numeric(
    df_new["price(in_dollar)"],
    errors="coerce"
)

df_new["raiting"] = pd.to_numeric(
    df_new["raiting"],
    errors="coerce"
)

df_new["reviews"] = pd.to_numeric(
    df_new["reviews"],
    errors="coerce"
)

In [ ]:
df_new.columns.tolist()

['price(in_dollar)', 'header', 'state', 'country', 'raiting', 'reviews']

In [ ]:
df_new["header"].head(20).tolist()

['Chalet in Skykomish',
 'Cabin in Hancock',
 'Cabin in West Farmington',
 'Home in Blue Ridge',
 'Treehouse in Grandview',
 'Tiny home in Puerto Escondido',
 'Home in Saco',
 'Chalet in Grand Étang',
 'Room in Mexico City',
 'Tiny home in Sandy Valley',
 'Guest suite in Huntsville',
 'Cabin in Nelson',
 'Dome in Joshua Tree',
 'Barn in Vero Beach',
 'Tiny home in Bolton',
 'Cabin in Ucluelet',
 'Barn in Essex',
 'Apartment in Bacalar',
 'Treehouse in Travelers Rest',
 'Farm stay in Del Norte County']

Finding location of every listing: We can use 'in' to get the location

In [ ]:
def header_extractor(df: pd.DataFrame) -> pd.DataFrame:
    temp_df = df.copy()

    # Split header into type and location
    split_header = temp_df["header"].str.split(" in ", n=1, expand=True)

    # Give the two resulting columns names
    split_header.columns = ["type", "location"]

    # Add them to the dataframe
    temp_df[["type", "location"]] = split_header

    # Remove original header
    temp_df.drop(columns=["header"], inplace=True)

    return temp_df

In [ ]:
df_location = header_extractor(df_new)
display(df_location.head(10))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
6,607.0,Maine,US,4.64,90.0,Home,Saco
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley


In [ ]:
df_location_copy=df_location.copy()

Airbnb data
     ↓
unique locations
     ↓
geocoding
     ↓
latitude + longitude
     ↓
JOIN back to Airbnb data??

In [ ]:
df_location_copy.isna().sum()

,0
price(in_dollar),1
state,235
country,733
raiting,20
reviews,20
type,0
location,0


In [ ]:
df_location_copy[df_location_copy["country"].isna()]

,price(in_dollar),state,country,raiting,reviews,type,location
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
10,207.0,Canada,None,4.97,265.0,Guest suite,Huntsville
11,162.0,Canada,None,4.99,226.0,Cabin,Nelson
...,...,...,...,...,...,...,...
948,141.0,None,None,4.85,82.0,Guest suite,Lake Worth
949,43.0,None,None,4.90,300.0,Room,Kissimmee
950,73.0,None,None,4.92,282.0,Guesthouse,San Antonio
951,233.0,None,None,4.96,318.0,Apartment,Englewood



**IMPORTANTE:** IN THE PAST CELL I FOUND OUT THAT THERE ARE ACTUALLY 733 NONE(NULL) VALUES IN THE COLLUMN COUNTRY
---



So it is really necessary to fill those values, using the information from the states, but to address the problem in a doable way, I will initially work with a subdata frame for the US

In [ ]:
us_states = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California",
    "Colorado", "Connecticut", "Delaware", "Florida", "Georgia",
    "Hawaii", "Idaho", "Illinois", "Indiana", "Iowa", "Kansas",
    "Kentucky", "Louisiana", "Maine", "Maryland", "Massachusetts",
    "Michigan", "Minnesota", "Mississippi", "Missouri", "Montana",
    "Nebraska", "Nevada", "New Hampshire", "New Jersey", "New Mexico",
    "New York", "North Carolina", "North Dakota", "Ohio", "Oklahoma",
    "Oregon", "Pennsylvania", "Rhode Island", "South Carolina",
    "South Dakota", "Tennessee", "Texas", "Utah", "Vermont",
    "Virginia", "Washington", "West Virginia", "Wisconsin", "Wyoming"
]

In [ ]:
mask = (
    df_location_copy["country"].isna()
    & df_location_copy["state"].isin(us_states)
)

df_location_copy.loc[mask, "country"] = "US"
display(df_location_copy.head(30))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
6,607.0,Maine,US,4.64,90.0,Home,Saco
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley


In [ ]:
df_location_copy.dropna(subset=["state"])
display(df_location_copy.head(30))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
6,607.0,Maine,US,4.64,90.0,Home,Saco
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley


In [ ]:
print(df_location_copy["country"].unique())

[' US' None ' Mexico' ' Italy' ' Thailand' ' Indonesia' ' Philippines'
 ' Malaysia']


In [ ]:
us_locations = df_location_copy[
    df_location_copy["country"] == " US"
][["location", "state"]]
display(us_locations.head(30))

,location,state
0,Skykomish,Washington
1,Hancock,New York
2,West Farmington,Ohio
3,Blue Ridge,Georgia
4,Grandview,Texas
6,Saco,Maine
9,Sandy Valley,Nevada
12,Joshua Tree,California
13,Vero Beach,Florida
14,Bolton,Vermont


In [ ]:
df_location_copy.to_csv("df_location_copy.csv", index=False)
df_location_copy = pd.read_csv("df_location_copy.csv")

In [ ]:
df_location_us = df_location_copy[df_location_copy["country"] == " US"].drop_duplicates()
display(df_location_us.head(10))
print(df_location_us.shape)

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
6,607.0,Maine,US,4.64,90.0,Home,Saco
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley
12,268.0,California,US,4.94,944.0,Dome,Joshua Tree
13,111.0,Florida,US,4.95,417.0,Barn,Vero Beach
14,174.0,Vermont,US,4.97,372.0,Tiny home,Bolton


(108, 7)


Here we safe our full data frame with location, rating, reviews, type, price, state and country

In [ ]:
#From now on the full data clean frame for the US is store in df_location_complete_us
df_location_complete_us = df_location_us.copy()
display(df_location_complete_us.head(10))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
6,607.0,Maine,US,4.64,90.0,Home,Saco
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley
12,268.0,California,US,4.94,944.0,Dome,Joshua Tree
13,111.0,Florida,US,4.95,417.0,Barn,Vero Beach
14,174.0,Vermont,US,4.97,372.0,Tiny home,Bolton


The problem **up here ** is that is not possible to extract the coordenates from the **Google Maps** links with the current level of coding and data analytics, so for this student project we will try Geopy instead


In [ ]:
df_location_us = (
    df_location_copy[
        df_location_copy["country"] == " US"
    ][["location", "state", "country"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

print("Unique US locations:", len(df_location_us))

display(df_location_us.head(20))

Unique US locations: 87


,location,state,country
0,Skykomish,Washington,US
1,Hancock,New York,US
2,West Farmington,Ohio,US
3,Blue Ridge,Georgia,US
4,Grandview,Texas,US
5,Saco,Maine,US
6,Sandy Valley,Nevada,US
7,Joshua Tree,California,US
8,Vero Beach,Florida,US
9,Bolton,Vermont,US


In [ ]:
import numpy as np

df_location_us["latitude"] = np.nan
df_location_us["longitude"] = np.nan

In [ ]:
display(df_location_us.head())

,location,state,country,latitude,longitude
0,Skykomish,Washington,US,NaN,NaN
1,Hancock,New York,US,NaN,NaN
2,West Farmington,Ohio,US,NaN,NaN
3,Blue Ridge,Georgia,US,NaN,NaN
4,Grandview,Texas,US,NaN,NaN


We are now getting the coordenates of the locations using geopy

In [ ]:
!pip install geopy

In [ ]:
from geopy.geocoders import Nominatim
from time import sleep

geolocator = Nominatim(
    user_agent="airbnb_analysis_project"
)

In [ ]:
def get_coordinates(row):

    address = f"{row['location']}, {row['state']}, {row['country']}"

    try:
        location = geolocator.geocode(
            address,
            timeout=10
        )

        if location:
            return pd.Series({
                "latitude": location.latitude,
                "longitude": location.longitude
            })

    except Exception as e:
        print(f"Error for {address}: {e}")

    return pd.Series({
        "latitude": np.nan,
        "longitude": np.nan
    })

Iterating to fill information of latitude and longitude in current data frame

In [ ]:
for index, row in df_location_us.iterrows():

    print(
        index,
        row["location"],
        row["state"]
    )

    coordinates = get_coordinates(row)

    df_location_us.loc[
        index, "latitude"
    ] = coordinates["latitude"]

    df_location_us.loc[
        index, "longitude"
    ] = coordinates["longitude"]

    sleep(1)

0 Skykomish  Washington
1 Hancock  New York
2 West Farmington  Ohio
3 Blue Ridge  Georgia
4 Grandview  Texas
5 Saco  Maine
6 Sandy Valley  Nevada
7 Joshua Tree  California
8 Vero Beach  Florida
9 Bolton  Vermont
10 Travelers Rest  South Carolina
11 Del Norte County  California
12 Frazeysburg  Ohio
13 Terlingua  Texas
14 Port Angeles  Washington
15 Fredericksburg  Texas
16 Bryson City  North Carolina
17 Marion  North Carolina
18 Sandgate  Vermont
19 Sanbornton  New Hampshire
20 Moss Beach  California
21 Topanga  California
22 Center Conway  New Hampshire
23 Huntsville  Missouri
24 Bodega Bay  California
25 Key Largo  Florida
26 Malibu  California
27 Jarratt  Virginia
28 Miami  Florida
29 Kansas City  Missouri
30 Island Park  Idaho
31 Arrowbear Lake  California
32 Shaftsbury  Vermont
33 Myrtle Beach  South Carolina
34 Laguna Beach  California
35 Lahaina  Hawaii
36 Hollywood  Florida
37 Naples  New York
38 Hilbert  Wisconsin
39 Whittier  North Carolina
40 Key West  Florida
41 Destin  Flor

In [ ]:
print(
    "Successfully geocoded:",
    df_location_us["latitude"].notna().sum()
)

print(
    "Failed:",
    df_location_us["latitude"].isna().sum()
)

Successfully geocoded: 87
Failed: 0


In [ ]:
display(df_location_us.sample(10))

,location,state,country,latitude,longitude
72,Naples,Florida,US,26.142198,-81.794294
82,Kissimmee,Florida,US,28.291899,-81.407584
2,West Farmington,Ohio,US,41.390499,-80.973621
13,Terlingua,Texas,US,29.321586,-103.616019
86,Wilmington,New York,US,44.388409,-73.815422
17,Marion,North Carolina,US,35.684102,-82.009103
25,Key Largo,Florida,US,25.087035,-80.447239
16,Bryson City,North Carolina,US,35.430929,-83.447381
23,Huntsville,Missouri,US,39.440589,-92.545183
35,Lahaina,Hawaii,US,20.887177,-156.674802


In [ ]:
df_location_us.to_csv("df_location_us.csv", index=False)
df_location_us_cordenates = df_location_us.copy()

NOW THAT WE HAVE A DATA FRAME WITH THE COORDINATES FOLIUM IS GOING TO WORK!

In [ ]:
import folium

m = folium.Map(
    location=[39.8283, -98.5795],
    zoom_start=4
)

In [ ]:
print(
    "Successfully geocoded:",
    df_location_us["latitude"].notna().sum()
)

print(
    "Failed:",
    df_location_us["latitude"].isna().sum()
)

Successfully geocoded: 87
Failed: 0


In [ ]:
display(df_location_us.head(20))

,location,state,country,latitude,longitude
0,Skykomish,Washington,US,47.709272,-121.360568
1,Hancock,New York,US,41.954223,-75.280540
2,West Farmington,Ohio,US,41.390499,-80.973621
3,Blue Ridge,Georgia,US,34.863972,-84.324087
4,Grandview,Texas,US,32.266651,-97.178543
5,Saco,Maine,US,43.500739,-70.442986
6,Sandy Valley,Nevada,US,35.830399,-115.627094
7,Joshua Tree,California,US,34.123375,-116.312866
8,Vero Beach,Florida,US,27.638716,-80.397540
9,Bolton,Vermont,US,44.373306,-72.878509


In [451]:
import pandas as pd

new_df = pd.merge(
    df_location_complete_us,
    df_location_us,
    on=["location", "state", "country"],
    how="left"
)
display(new_df.head(10))

,price(in_dollar),state,country,raiting,reviews,type,location,latitude,longitude
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish,47.709272,-121.360568
1,485.0,New York,US,4.77,146.0,Cabin,Hancock,41.954223,-75.280540
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington,41.390499,-80.973621
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge,34.863972,-84.324087
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview,32.266651,-97.178543
5,607.0,Maine,US,4.64,90.0,Home,Saco,43.500739,-70.442986
6,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley,35.830399,-115.627094
7,268.0,California,US,4.94,944.0,Dome,Joshua Tree,34.123375,-116.312866
8,111.0,Florida,US,4.95,417.0,Barn,Vero Beach,27.638716,-80.397540
9,174.0,Vermont,US,4.97,372.0,Tiny home,Bolton,44.373306,-72.878509


In [454]:
df_location_complete_us = df_location_complete_us.merge(
    df_location_us[
        [
            "location",
            "state",
            "country",
            "latitude",
            "longitude"
        ]
    ],
    on=["location", "state", "country"],
    how="left"
)

In [455]:
df_geoLocalization=new_df.copy()
display(df_geoLocalization.head(10))

,price(in_dollar),state,country,raiting,reviews,type,location,latitude,longitude
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish,47.709272,-121.360568
1,485.0,New York,US,4.77,146.0,Cabin,Hancock,41.954223,-75.280540
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington,41.390499,-80.973621
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge,34.863972,-84.324087
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview,32.266651,-97.178543
5,607.0,Maine,US,4.64,90.0,Home,Saco,43.500739,-70.442986
6,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley,35.830399,-115.627094
7,268.0,California,US,4.94,944.0,Dome,Joshua Tree,34.123375,-116.312866
8,111.0,Florida,US,4.95,417.0,Barn,Vero Beach,27.638716,-80.397540
9,174.0,Vermont,US,4.97,372.0,Tiny home,Bolton,44.373306,-72.878509


In [456]:
df_location_complete_us.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   price(in_dollar)  107 non-null    float64
 1   state             108 non-null    object 
 2   country           108 non-null    object 
 3   raiting           102 non-null    float64
 4   reviews           102 non-null    float64
 5   type              108 non-null    object 
 6   location          108 non-null    object 
 7   latitude          108 non-null    float64
 8   longitude         108 non-null    float64
dtypes: float64(5), object(4)
memory usage: 7.7+ KB


In [ ]:
import folium

m = folium.Map(
    location=[39.8283, -98.5795],
    zoom_start=4
)

In [ ]:
map_df = new_df.dropna(
    subset=[
        "latitude",
        "longitude",
        "price(in_dollar)",
        "raiting"
    ]
).copy()

In [ ]:
print("Listings available for map:", len(map_df))

Listings available for map: 101


In [ ]:
display(
    map_df[
        [
            "location",
            "state",
            "latitude",
            "longitude",
            "price(in_dollar)",
            "raiting"
        ]
    ].head(10)
)

,location,state,latitude,longitude,price(in_dollar),raiting
0,Skykomish,Washington,47.709272,-121.360568,306.0,4.85
1,Hancock,New York,41.954223,-75.280540,485.0,4.77
2,West Farmington,Ohio,41.390499,-80.973621,119.0,4.91
3,Blue Ridge,Georgia,34.863972,-84.324087,192.0,4.94
4,Grandview,Texas,32.266651,-97.178543,232.0,4.99
5,Saco,Maine,43.500739,-70.442986,607.0,4.64
6,Sandy Valley,Nevada,35.830399,-115.627094,154.0,4.92
7,Joshua Tree,California,34.123375,-116.312866,268.0,4.94
8,Vero Beach,Florida,27.638716,-80.397540,111.0,4.95
9,Bolton,Vermont,44.373306,-72.878509,174.0,4.97


In [ ]:
import folium

m = folium.Map(
    location=[39.8283, -98.5795],
    zoom_start=4
)

In [ ]:
def rating_color(rating):
    if rating >= 4.8:
        return "green"
    elif rating >= 4.5:
        return "lightgreen"
    elif rating >= 4.0:
        return "orange"
    else:
        return "red"

In [ ]:
for _, row in map_df.iterrows():

    popup_text = f"""
    <h4>{row['type']}</h4>

    <b>Location:</b> {row['location']}<br>
    <b>State:</b> {row['state']}<br>
    <b>Price:</b> ${row['price(in_dollar)']:.0f}<br>
    <b>Rating:</b> {row['raiting']:.1f}<br>
    <b>Reviews:</b> {row['reviews']}<br><br>

    <a href="{row['google_maps']}" target="_blank">
        📍 Open in Google Maps
    </a>
    """

    folium.CircleMarker(
        location=[
            row["latitude"],
            row["longitude"]
        ],
        radius=price_size(row["price(in_dollar)"]),
        color=rating_color(row["raiting"]),
        fill=True,
        fill_color=rating_color(row["raiting"]),
        fill_opacity=0.7,
        popup=folium.Popup(
            popup_text,
            max_width=300
        )
    ).add_to(m)

m

KeyError: 'google_maps'